kv_test.go

```diff
@@ -195,6 +196,123 @@ func TestCompareAndDelete(t *testing.T) {
 	assert.True(t, deleted)
 }
 
+func TestSetAtomicWithRetries(t *testing.T) {
+	tests := []struct {
+		name      string
+		key       string
+		valueFunc func(oldValue interface{}) (newValue interface{}, err error)
+		setupAPI  func(api *plugintest.API)
+		wantErr   bool
+	}{
+		{
+			name: "Test SetAtomicWithRetries success after first attempt",
+			key:  "testNum",
+			valueFunc: func(old interface{}) (interface{}, error) {
+				return 2, nil
+			},
+			setupAPI: func(api *plugintest.API) {
+				oldJSONBytes, _ := json.Marshal(1)
+				newJSONBytes, _ := json.Marshal(2)
+				api.On("KVGet", "testNum").Return(oldJSONBytes, nil)
+				api.On("KVSetWithOptions", "testNum", newJSONBytes, model.PluginKVSetOptions{
+					Atomic:   true,
+					OldValue: oldJSONBytes,
+				}).Return(true, nil)
+			},
+		},
+		{
+			name: "Test SetAtomicWithRetries success on fourth attempt",
+			key:  "testNum",
+			valueFunc: func(old interface{}) (interface{}, error) {
+				return 2, nil
+			},
+			setupAPI: func(api *plugintest.API) {
+				oldJSONBytes, _ := json.Marshal(1)
+				newJSONBytes, _ := json.Marshal(2)
+				api.On("KVGet", "testNum").Return(oldJSONBytes, nil).Times(4)
+				api.On("KVSetWithOptions", "testNum", newJSONBytes, model.PluginKVSetOptions{
+					Atomic:   true,
+					OldValue: oldJSONBytes,
+				}).Return(false, nil).Times(3)
+				api.On("KVSetWithOptions", "testNum", newJSONBytes, model.PluginKVSetOptions{
+					Atomic:   true,
+					OldValue: oldJSONBytes,
+				}).Return(true, nil).Once()
+			},
+		},
+		{
+			name: "Test SetAtomicWithRetries failure on get",
+			key:  "testNum",
+			valueFunc: func(old interface{}) (interface{}, error) {
+				return 2, nil
```

Comment:
should we assert in this case that this func was not called?

Reply:
I'm not sure how that would be possible to do, and not sure how it would be useful...?

Reply:
Could update a flag in here and assert that. But maybe it's not that useful. I was thinking if the client is doing an external operation that has a side effect, calling this func erroneously (in this case) in those scenarios it may lead to problems. But since those cases will be rare maybe it's not worth the hassle. More a correctness concern than anything. 0/5 though if we should assert.

Reply:
Okay. No need for a flag, I'm testing the error returned, that will tell you where it failed.